File: src/wings/tools/builtin/grep.py
"""Regular expression content search."""

import re
from pathlib import Path

from pydantic import BaseModel, Field

from wings.tools.base import ToolContext
from wings.tools.decorator import tool


class GrepInput(BaseModel):
    pattern: str = Field(description="The regular expression pattern to search for")
    path: str | None = Field(
        default=None,
        description="File or directory to search in. Defaults to working directory.",
    )
    glob: str | None = Field(
        default=None,
        description="Glob pattern to filter files, e.g. '*.py'",
    )
    output_mode: str | None = Field(
        default="content",
        description="Output mode: 'content', 'files_with_matches', or 'count'",
    )
    head_limit: int | None = Field(
        default=None,
        description="Limit output to first N lines/entries",
    )


@tool(
    name="grep",
    description="Search file contents using regular expressions",
    search_hint="grep 'pattern'",
    read_only=True,
)
async def grep(input: GrepInput, context: ToolContext) -> str:
    base = Path(input.path) if input.path else Path(context.working_dir)
    if not base.is_absolute():
        base = Path(context.working_dir) / base
    if not base.exists():
        return f"Error: path not found: {base}"

    # Collect files to search
    if base.is_file():
        files = [base]
    else:
        glob_pattern = input.glob or "**/*"
        files = sorted(
            f for f in base.glob(glob_pattern) if f.is_file()
        )

    try:
        regex = re.compile(input.pattern)
    except re.error as e:
        return f"Error: invalid regex: {e}"

    output_lines: list[str] = []
    file_count = 0

    for filepath in files:
        try:
            text = filepath.read_text()
        except (UnicodeDecodeError, OSError):
            continue

        matches = list(regex.finditer(text))
        if not matches:
            continue

        file_count += 1

        if input.output_mode == "files_with_matches":
            output_lines.append(str(filepath))
        elif input.output_mode == "count":
            output_lines.append(f"{str(filepath)}:{len(matches)}")
        else:
            lines = text.split("\n")
            for m in matches:
                line_no = text[:m.start()].count("\n") + 1
                line_text = lines[line_no - 1] if line_no <= len(lines) else ""
                output_lines.append(f"{filepath}:{line_no}: {line_text.strip()}")

        if input.head_limit and len(output_lines) >= input.head_limit:
            output_lines = output_lines[:input.head_limit]
            break

    if not output_lines:
        return "(no matches)"

    if input.head_limit:
        output_lines = output_lines[:input.head_limit]

    return "\n".join(output_lines)

File: src/tools/builtin/grep.ts
/** Regular expression content search. */

import { existsSync, statSync, readFileSync, readdirSync } from "node:fs";
import { isAbsolute, join } from "node:path";

import { z } from "zod";

import { buildTool } from "../types.ts";

// Directories to exclude from search
const VCS_DIRS = new Set([".git", ".svn", ".hg", ".bzr", ".jj", ".sl"]);

interface GrepInput {
  pattern: string;
  path?: string | null;
  glob?: string | null;
  output_mode?: string | null;
  head_limit?: number | null;
}


/** Recursively collect files under base, optionally filtered by glob pattern. */
function collectFiles(base: string, globPattern: string | null): string[] {
  const result: string[] = [];
  function walk(dir: string) {
    let entries: import("node:fs").Dirent[];
    try {
      entries = readdirSync(dir, { withFileTypes: true }) as unknown as import("node:fs").Dirent[];
    } catch {
      return;
    }
    for (const entry of entries) {
      const name: string = entry.name as unknown as string;
      const full = join(dir, name);
      if (entry.isDirectory()) {
        if (!VCS_DIRS.has(name)) walk(full);
      } else if (entry.isFile()) {
        if (globPattern && globPattern !== "**/*") {
          if (!matchSimpleGlob(name, globPattern)) continue;
        }
        result.push(full);
      }
    }
  }
  walk(base);
  return result.sort();
}

function matchSimpleGlob(name: string, pattern: string): boolean {
  // Convert simple glob to regex: *, ?, {a,b}
  let regex = "^";
  let i = 0;
  while (i < pattern.length) {
    const c = pattern[i];
    if (c === "*") {
      regex += ".*";
    } else if (c === "?") {
      regex += ".";
    } else if (c === "{") {
      const end = pattern.indexOf("}", i);
      if (end === -1) {
        regex += "\\{";
      } else {
        const options = pattern.slice(i + 1, end).split(",").map((s) => s.trim());
        regex += `(?:${options.join("|")})`;
        i = end;
      }
    } else if (c === "[") {
      regex += "[";
    } else if (c === "]") {
      regex += "]";
    } else if ("\\^$.|+()".includes(c)) {
      regex += "\\" + c;
    } else {
      regex += c;
    }
    i++;
  }
  regex += "$";
  return new RegExp(regex).test(name);
}

export const grepTool = buildTool({
  name: "grep",
  description: "Search file contents using regular expressions",
  search_hint: "grep 'pattern'",
  is_read_only: true,
  inputSchema: z.object({
    pattern: z.string().describe("The regular expression pattern to search for"),
    path: z.string().optional().describe("File or directory to search in. Defaults to working directory."),
    glob: z.string().optional().describe("Glob pattern to filter files, e.g. '*.py'"),
    output_mode: z.string().optional().describe("Output mode: 'content', 'files_with_matches', or 'count'"),
    head_limit: z.number().optional().describe("Limit output to first N lines/entries"),
  }),
  async call(input: GrepInput, context) {
    let base = input.path ?? context.working_dir;
    if (!isAbsolute(base)) base = join(context.working_dir, base);
    if (!existsSync(base)) {
      return `Error: path not found: ${base}`;
    }

    // Collect files to search
    let files: string[];
    const stat = statSync(base);
    if (stat.isFile()) {
      files = [base];
    } else {
      files = collectFiles(base, input.glob ?? null);
    }

    let regex: RegExp;
    try {
      regex = new RegExp(input.pattern);
    } catch (e) {
      return `Error: invalid regex: ${(e as Error).message}`;
    }

    const outputLines: string[] = [];
    let fileCount = 0;
    let matchCount = 0;
    const outputMode = input.output_mode ?? "content";

    for (const filepath of files) {
      let text: string;
      try {
        text = readFileSync(filepath, "utf-8");
      } catch {
        continue;
      }

      // Find all matches. Strip sticky flag (y) and ensure global (g) for exec loop.
      let baseFlags = regex.flags.replace("y", "");
      if (!baseFlags.includes("g")) baseFlags += "g";
      const globalRegex = new RegExp(regex.source, baseFlags);
      const matches: Array<{ index: number }> = [];
      let m: RegExpExecArray | null;
      while ((m = globalRegex.exec(text)) !== null) {
        matches.push({ index: m.index });
        if (m.index === globalRegex.lastIndex) globalRegex.lastIndex++;
      }

      if (matches.length === 0) continue;

      fileCount++;
      matchCount += matches.length;

      if (outputMode === "files_with_matches") {
        outputLines.push(filepath);
      } else if (outputMode === "count") {
        outputLines.push(`${filepath}:${matches.length}`);
      } else {
        const lines = text.split("\n");
        for (const match of matches) {
          const lineNo = text.slice(0, match.index).split("\n").length;
          const lineText = lineNo <= lines.length ? lines[lineNo - 1] : "";
          outputLines.push(`${filepath}:${lineNo}: ${lineText.trim()}`);
        }
      }

      if (input.head_limit && outputLines.length >= input.head_limit) {
        outputLines.length = input.head_limit;
        break;
      }
    }

    if (outputLines.length === 0) {
      return "(no matches)";
    }

    // Summary line matching claude-code: "Found N matches across M files"
    let summary: string;
    if (outputMode === "files_with_matches") {
      const label = fileCount === 1 ? "file" : "files";
      summary = `Found ${fileCount} ${label}`;
    } else if (outputMode === "count") {
      const label = matchCount === 1 ? "match" : "matches";
      const fileLabel = fileCount === 1 ? "file" : "files";
      summary = `Found ${matchCount} ${label} across ${fileCount} ${fileLabel}`;
    } else {
      const label = matchCount === 1 ? "match" : "matches";
      const fileLabel = fileCount === 1 ? "file" : "files";
      summary = `Found ${matchCount} ${label} across ${fileCount} ${fileLabel}`;
    }

    return summary + "\n" + outputLines.join("\n");
  },
});
